### PR TITLE
Fix translog replay multiple operations same doc

### PR DESCRIPTION
**Pull request by jasontedor**
Modifying the translog replay to not replay again into the translog
introduced a bug for the case of multiple operations for the same
doc. Namely, since we were no longer updating the version map for each
operation, the second operation for a doc would be treated as a creation
instead of as an update. This commit fixes this bug by placing these
operations into version map. This commit includes a failing test case.

Relates #18547
Relates #18623

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 1]
```diff
@@ -591,6 +592,42 @@ public void testFlushIsDisabledDuringTranslogRecovery() throws IOException {
         engine.flush();
     }
 
+    public void testTranslogMultipleOperationsSameDocument() throws IOException {
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: good!!

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -394,6 +394,8 @@ private boolean innerIndex(Index index) throws IOException {
                 final Translog.Location translogLocation = translog.add(new Translog.Index(index));
                 index.setTranslogLocation(translogLocation);
                 versionMap.putUnderLock(index.uid().bytes(), new VersionValue(updatedVersion, index.getTranslogLocation()));
+            } else {
+                versionMap.putUnderLock(index.uid().bytes(), new VersionValue(updatedVersion, null));
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can you put a comment here that we flush the version map after recovery so null is fine here...

**Comment by s1monw** [comment]
LGMT left a small comment

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I pushed 41a953b3fb086d402123573c9b93bc07791649cd.

**Pull request merged**